transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -872,12 +892,90 @@ private AbstractWriteTask(Recycler.Handle handle) {
             super(handle);
         }
 
+        private static int estimateSize(Object o, Map<Class<?>, Integer> classSizes) {
+            int answer = 8 + estimateSize(o.getClass(), classSizes, null);
+
+            if (o instanceof ByteBuf) {
+                answer += ((ByteBuf) o).readableBytes();
+            } else if (o instanceof ByteBufHolder) {
+                answer += ((ByteBufHolder) o).content().readableBytes();
+            } else if (o instanceof FileRegion) {
+                // nothing to add.
+            } else if (o instanceof byte[]) {
+                answer += ((byte[]) o).length;
+            } else if (o instanceof ByteBuffer) {
+                answer += ((ByteBuffer) o).remaining();
+            } else if (o instanceof CharSequence) {
+                answer += ((CharSequence) o).length() << 1;
+            } else if (o instanceof Iterable<?>) {
+                for (Object m : (Iterable<?>) o) {
+                    answer += estimateSize(m, classSizes);
+                }
+            }
+
+            return align(answer);
+        }
+
+        private static int estimateSize(Class<?> clazz, Map<Class<?>, Integer> classSizes,
+                                        Set<Class<?>> visitedClasses) {
+            Integer objectSize = classSizes.get(clazz);
+            if (objectSize != null) {
+                return objectSize;
+            }
+
+            if (visitedClasses != null) {
+                if (visitedClasses.contains(clazz)) {
+                    return 0;
+                }
+            } else {
+                visitedClasses = new HashSet<Class<?>>();
+            }
+
+            visitedClasses.add(clazz);
+
+            int answer = 8; // Basic overhead.
+            for (Class<?> c = clazz; c != null; c = c.getSuperclass()) {
+                Field[] fields = c.getDeclaredFields();
+                for (Field f : fields) {
+                    if ((f.getModifiers() & Modifier.STATIC) != 0) {
+                        // Ignore static fields.
+                        continue;
+                    }
+
+                    answer += estimateSize(f.getType(), classSizes, visitedClasses);
+                }
+            }
+
+            visitedClasses.remove(clazz);
+
+            // Some alignment.
+            answer = align(answer);
+
+            // Put the final answer.
+            classSizes.put(clazz, answer);
+            return answer;
+        }
+
+        private static int align(int size) {
+            int r = size & 7;
+            if (r != 0) {
```

Comment:
I wonder if we need this conditional? Maybe just remove the conditional and do the simple arithmetic...maybe cheaper than doing the conditional.

```java
@Param({ "10" })
public int size;

@Benchmark
@BenchmarkMode(Mode.Throughput)
public int withConditional() {
  int r = size & 7;
  if (r != 0) {
     size += 8 - r;
  }
  return size;
}

@Benchmark
@BenchmarkMode(Mode.Throughput)
public int noConditional() {
  return size + 8 - (size & 7);
}
```

```
NetUtilBenchmark.noConditional        10  thrpt   20  1537030516.089 ± 26026477.373  ops/s
NetUtilBenchmark.withConditional      10  thrpt   20  1500392559.217 ± 20979325.309  ops/s
```